### PR TITLE
.Net IKernel Function Hooks Phase 2 (Rendered Prompt Support)

### PR DESCRIPTION
**Pull request by RogerBarreto**
### Motivation and Context

Currently Kernel invoking and invoked handlers don't expose the prompt to the handlers.

The proposal is a way to expose the prompt to the handlers.

- Pre-Execution / Invoking

  - Get: Prompt generated by the current SemanticFunction.TemplateEngine before calling the LLM
  - Set: Modify a prompt content before sending it to LLM
  - Post-Execution / Invoked

- Post-Execution / Invoked
  - Get: Generated Prompt

### Description

This PR has the implementation of the decided option 4 in the ADR also present as one of the PR documents.

Below is the expected end user experience when coding using Pre/Post Hooks to get or modify prompts.

### Extras
  - Added missing capability of cancelling a Plan from a step execution.
  - Added missing UT's regarding more handling scenarios in plans and kernel.
  - Added one Integration Test on handling scenarios with hooks in plan steps.

```csharp
const string FunctionPrompt = "Write a random paragraph about: {{$input}}.";

var excuseFunction = kernel.CreateSemanticFunction(...);

void MyPreHandler(object? sender, FunctionInvokingEventArgs e)
{
    Console.WriteLine($"{e.FunctionView.PluginName}.{e.FunctionView.Name} : Pre Execution Handler - Triggered");

    // Will be false for non semantic functions
    if (e.TryGetRenderedPrompt(out var prompt))
    {
        Console.WriteLine("Rendered Prompt:");
        Console.WriteLine(prompt);

        // Update the prompt if needed
        e.TryUpdateRenderedPrompt("Write a random paragraph about: Overriding a prompt");
    }
}

void MyPostHandler(object? sender, FunctionInvokedEventArgs e)
{
    Console.WriteLine($"{e.FunctionView.PluginName}.{e.FunctionView.Name} : Post Execution Handler - Triggered");
    // Will be false for non semantic functions
    if (e.TryGetRenderedPrompt(out var prompt))
    {
        Console.WriteLine("Used Prompt:");
        Console.WriteLine(prompt);
    }
}

kernel.FunctionInvoking += MyPreHandler;
kernel.FunctionInvoked += MyPostHandler;

const string Input = "I missed the F1 final race";
var result = await kernel.RunAsync(Input, excuseFunction);
Console.WriteLine($"Function Result: {result.GetValue<string>()}");
```

Expected output:

```
MyPlugin.MyFunction : Pre Execution Handler - Triggered
Rendered Prompt:
Write a random paragraph about: I missed the F1 final race.

MyPlugin.MyFunction : Post Execution Handler - Triggered
Used Prompt:
Write a random paragraph about: Overriding a prompt

FunctionResut: <LLM Completion>
```
### Contribution Checklist

- [x] The code builds clean without any errors or warnings
- [x] The PR follows the [SK Contribution Guidelines](https://github.com/microsoft/semantic-kernel/blob/main/CONTRIBUTING.md) and the [pre-submission formatting script](https://github.com/microsoft/semantic-kernel/blob/main/CONTRIBUTING.md#development-scripts) raises no violations
- [x] All unit tests pass, and I have added new tests where possible
- [ ] I didn't break anyone :smile:

### CODE REVIEW

File: dotnet/src/SemanticKernel.Abstractions/Functions/EventDelegateWrapper{TEventArgs}.cs [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+
+#pragma warning disable IDE0130
+// ReSharper disable once CheckNamespace - Using the main namespace
+namespace Microsoft.SemanticKernel;
+
+/// <summary>
+/// Class for storing event handler and event args for function events.
+/// </summary>
+/// <typeparam name="TEventArgs"></typeparam>
+public sealed class EventHandlerWrapper<TEventArgs> where TEventArgs : SKEventArgs
```
Comment (dmytrostruk) [inline dotnet/src/SemanticKernel.Abstractions/Functions/EventDelegateWrapper{TEventArgs}.cs:14]: Class name doesn't match file name.
Class name: `EventHandlerWrapper<TEventArgs>`.
File name: `EventDelegateWrapper{TEventArgs}.cs`

**Review by dmytrostruk — COMMENTED** [review]
@RogerBarreto thank you for the great work! 
I added initial comments with proposals how we can simplify current implementation. Please let me know what you think about it.

Also, as part of this PR, it would be nice to see unit and/or integration tests to check that rendered prompt is accessible from hooks and can be modified before sending request to LLM.

Thanks again!

File: dotnet/src/SemanticKernel.Core/Kernel.cs [hunk 2]
```diff
@@ -120,40 +121,31 @@ public async Task<KernelResult> RunAsync(ContextVariables variables, Cancellatio
...
+                var invokingEventWrapper = new EventHandlerWrapper<FunctionInvokingEventArgs>(this.FunctionInvoking);
+                var invokedEventWrapper = new EventHandlerWrapper<FunctionInvokedEventArgs>(this.FunctionInvoked);
 
-                context = functionResult.Context;
+                functionResult = await skFunction.InvokeAsync(context, null, invokingEventWrapper, invokedEventWrapper, cancellationToken: cancellationToken).ConfigureAwait(false);
 
-                var functionInvokedArgs = this.OnFunctionInvoked(functionDetails, functionResult);
-
-                if (functionInvokedArgs is not null)
+                if (functionResult is StopFunctionResult stopResult)
```
[the remark this answers is not in the document]
Comment (RogerBarreto) [reply dotnet/src/SemanticKernel.Core/Kernel.cs]: In my last commit I removed the StopFunctionResult entirely, thanks for the feedback.

File: dotnet/src/SemanticKernel.Abstractions/Functions/EventDelegateWrapper{TEventArgs}.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (RogerBarreto) [reply dotnet/src/SemanticKernel.Abstractions/Functions/EventDelegateWrapper{TEventArgs}.cs]: As the logic changed from getting the eventWrapper from an argument to the SKContext, a lot of it was simplified from the InvokeAsync method impl.

Comment (RogerBarreto) [reply dotnet/src/SemanticKernel.Abstractions/Functions/EventDelegateWrapper{TEventArgs}.cs]: We can't have file names with `<>`, Renaming it to just the `EventHandlerWrapperOfT` as the one of the approach used in the [dotnet runtime repo](https://github.com/dotnet/runtime/blob/main/src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonConverterOfT.cs).

File: dotnet/src/SemanticKernel.Abstractions/Functions/EventDelegateWrapper{TEventArgs}.cs [hunk 3]
```diff
@@ -0,0 +1,34 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+
+using System;
+using Microsoft.SemanticKernel.Events;
+
+#pragma warning disable IDE0130
+// ReSharper disable once CheckNamespace - Using the main namespace
+namespace Microsoft.SemanticKernel;
```
[the remark this answers is not in the document]
Comment (RogerBarreto) [reply dotnet/src/SemanticKernel.Abstractions/Functions/EventDelegateWrapper{TEventArgs}.cs:8]: This pragma needs to come before the namespace as its complaining about the namespace not being the same as the folder.

File: dotnet/src/SemanticKernel.Abstractions/Functions/EventHandlerWrapper.cs [hunk 4]
```diff
@@ -0,0 +1,12 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+
+#pragma warning disable IDE0130
+// ReSharper disable once CheckNamespace - Using the main namespace
+namespace Microsoft.SemanticKernel;
+
+/// <summary>
+/// Flag abstract class for storing event handler and event args for function events.
+/// </summary>
+public abstract class EventHandlerWrapper
```
Comment (markwallace-microsoft) [inline dotnet/src/SemanticKernel.Abstractions/Functions/EventHandlerWrapper.cs:10]: An empty abstract class is a code smell. We have been working to eliminate empty interfaces. Why is this needed?

At a minimum we should explain this in the class documentation.

File: dotnet/src/SemanticKernel.Abstractions/Orchestration/SKContext.cs [hunk 5]
```diff
@@ -97,6 +110,28 @@ public CultureInfo Culture
...
+                if (handler is EventHandlerWrapper<FunctionInvokingEventArgs> invokingWrapper)
+                {
+                    this.FunctionInvokingHandler = invokingWrapper;
+                    continue;
+                }
+
+                if (handler is EventHandlerWrapper<FunctionInvokedEventArgs> invokedWrapper)
+                {
+                    this.FunctionInvokedHandler = invokedWrapper;
+                }
```
Comment (markwallace-microsoft) [inline dotnet/src/SemanticKernel.Abstractions/Orchestration/SKContext.cs:132]: If it's not one of the two types listed we show throw an error

File: dotnet/src/SemanticKernel.Core/Events/SKEventArgsExtensions.cs [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+{
+    /// <summary>
+    /// Only tries to get the rendered prompt from the event args metadata if it exists.
+    /// </summary>
+    /// <param name="eventArgs">Target event args to extend</param>
+    /// <param name="renderedPrompt">Outputs the renderedPrompt</param>
+    /// <returns>True if the prompt was present</returns>
+    public static bool TryGetRenderedPrompt(this SKEventArgs eventArgs, out string? renderedPrompt)
+    {
+        var found = eventArgs.Metadata.TryGetValue(SemanticFunction.RenderedPromptMetadataKey, out var renderedPromptObject);
```
Comment (markwallace-microsoft) [inline dotnet/src/SemanticKernel.Core/Events/SKEventArgsExtensions.cs:20]: Is `SemanticFunction` the right place to store the key name? That class is internal so I want to write my own function to pull values out of the Metadata I don't have an easy way to discover the keys.

File: dotnet/src/SemanticKernel.Core/Functions/NativeFunction.cs [hunk 7]
```diff
@@ -138,7 +139,21 @@ public FunctionView Describe()
...
-            return await this._function(null, requestSettings, context, cancellationToken).ConfigureAwait(false);
+            this.CallFunctionInvoking(context);
+            if (SKFunction.IsInvokingCancelOrSkipRequested(context))
+            {
+                return new FunctionResult(this.Name, this.PluginName, context);
+            }
+
+            var invokeResult = await this._function(null, requestSettings, context, cancellationToken).ConfigureAwait(false);
+            var finalResult = this.CallFunctionInvoked(invokeResult, context);
+
```
Comment (markwallace-microsoft) [inline dotnet/src/SemanticKernel.Core/Functions/NativeFunction.cs:150]: Move empty line to 149, males more sense to keep the `CallFunctionInvoked` beside the `if` check

File: dotnet/src/SemanticKernel.Core/Functions/SemanticFunction.cs [hunk 8]
```diff
@@ -204,6 +222,72 @@ private void AddDefaultValues(ContextVariables variables)
...
+        var eventWrapper = context.FunctionInvokingHandler;
+        if (eventWrapper?.Handler is null)
+        {
+            return;
+        }
+
+        eventWrapper.EventArgs = new FunctionInvokingEventArgs(this.Describe(), context)
+        {
+            Metadata = {
+                [SemanticFunction.RenderedPromptMetadataKey] = renderedPrompt
```
Comment (markwallace-microsoft) [inline dotnet/src/SemanticKernel.Core/Functions/SemanticFunction.cs:241]: Need to think about the discoverability of the metadata. Keys should be defined somewhere more obvious.

**Comment by SergeyMenshykh** [comment]
The PR adds two new properties to the `SKContext` class. Considering that we are still deciding what to do with the context, I would propose to postpone the PR until the decision is made in order to avoid reimplementing the hooks again if it's decided to abandon the class or repurpose it.

File: dotnet/src/SemanticKernel.Core/Events/SKEventArgsExtensions.cs (hunk 6 above)
Comment (RogerBarreto) [reply dotnet/src/SemanticKernel.Core/Events/SKEventArgsExtensions.cs:20]: I see, one way would be to add that constant to the SKEventArgs class itself, as this applies to the Metadata dictionary of it. What you think?

Comment (RogerBarreto) [reply dotnet/src/SemanticKernel.Core/Events/SKEventArgsExtensions.cs:20]: Agreed in moving this to SkEventArgs class.

File: dotnet/src/SemanticKernel.Abstractions/Functions/EventHandlerWrapper.cs (hunk 4 above)
Comment (RogerBarreto) [reply dotnet/src/SemanticKernel.Abstractions/Functions/EventHandlerWrapper.cs:10]: Agree, will change to code to avoid this.

File: dotnet/samples/KernelSyntaxExamples/Example57_FunctionEventHandlers.cs [hunk 9]
```diff
@@ -91,6 +93,59 @@ void MyPostExecutionHandler(object? sender, FunctionInvokedEventArgs e)
...
+            FunctionPrompt,
+            pluginName: "MyPlugin",
+            functionName: "Excuse",
+            requestSettings: new OpenAIRequestSettings() { MaxTokens = 100, Temperature = 0.4, TopP = 1 });
+
+        void MyPreHandler(object? sender, FunctionInvokingEventArgs e)
+        {
+            Console.WriteLine($"{e.FunctionView.PluginName}.{e.FunctionView.Name} : Pre Execution Handler - Triggered");
+
+            if (e.TryGetRenderedPrompt(out var prompt))
```
[the remark this answers is not in the document]
Comment (RogerBarreto) [reply dotnet/samples/KernelSyntaxExamples/Example57_FunctionEventHandlers.cs:119]: We could potentilly pass any data to Metadata dictionary, I'm not passing anything extra as this would be Out of Scope of this PR. We can add those once the LLM Multi Model/Model support is in place with the new IAIService interfaces.

File: dotnet/samples/KernelSyntaxExamples/Example57_FunctionEventHandlers.cs [hunk 10]
```diff
@@ -32,6 +32,8 @@ public static async Task RunAsync()
 
         await GetUsageAsync();
 
+        await GetRenderedPromptAsync();
```
[the remark this answers is not in the document]
Comment (RogerBarreto) [reply dotnet/samples/KernelSyntaxExamples/Example57_FunctionEventHandlers.cs:35]: Good point, I renamed the class as all the methods actually are related to hooks, new name `Example57_KernelHooks.cs`

**Review by markwallace-microsoft — APPROVED** [review]

**Review by SergeyMenshykh — APPROVED** [review]
